[9076] Remove CE chunked header from t.w.http CONNECT response

izzy opened the pull request:
According to [RFC 7231 Section 4.3.6](https://tools.ietf.org/html/rfc7231#page-31) "A server MUST NOT send any Transfer-Encoding or Content-Length header fields in a 2xx (Successful) response to CONNECT"

Fix for [Trac ticket 9076](https://twistedmatrix.com/trac/ticket/9076#ticket)

mkzero committed:
Remove CE chunked header from t.w.http response

According to RFC 7231 Section 4.3.6[1] "A server MUST NOT send
any Transfer-Encoding or Content-Length header fields in a
2xx (Successful) response to CONNECT"

  [1]: https://tools.ietf.org/html/rfc7231#page-31

exarkun commented:
Thanks.  Please add test coverage and a news fragment.

izzy commented:
Are there any resources on how to add those tests? I'm new to developing twisted itself and kinda lost here what to do and how to do it. Especially got a problem with the fact that twisted does not fully support CONNECT requests. Any input on how to get started with this?

izzy commented:
It seems that codecov report is a "little" off? Well anyway, added a test, added the topfile, as far as I can tell that should be it.

exarkun commented:
> It seems that codecov report is a "little" off? Well anyway, added a test, added the topfile, as far as I can tell that should be it.

Yea, the codecov reports aren't really useful because there is so much random variation.  Sorry about that.  Thanks for adding the test - sorry I didn't get a chance to give you any hints on that.

glyph commented:
While codecov definitely has accuracy issues, in _this_ case it's just reporting a lower number because it reports combined coverage, and no project member has pushed this change to a branch so the buildbots can run the full suite yet.  (This is also why the OS X builder is showing an unfinished status).

markrwilliams committed:
Merge branch 'trunk' into 9076-mkzero-t-w-web-connect-compatibility

hawkowl committed:
Merge branch 'trunk' into 9076-mkzero-t-w-web-connect-compatibility

njsmith commented on src/twisted/web/http.py:
```diff
@@ -1035,6 +1039,10 @@ def write(self, data):
                         value = networkString('%s' % (value,))
+                    if self.method == b"CONNECT" and name == b"Content-Length":
+                        log.msg("Warning: Removing Content-Length header in"
+                                " response to CONNECT request")
```
`Content-Length` is totally unobjectionable in, say, a 4xx response to a CONNECT. It's only 2xx responses that are special.

njsmith commented on src/twisted/web/http.py:
```diff
@@ -1008,9 +1008,13 @@ def write(self, data):
                 (self.responseHeaders.getRawHeaders(b'content-length') is None) and
-                self.method != b"HEAD" and self.code not in NO_BODY_CODES):
+                self.method != b"HEAD" and self.code not in NO_BODY_CODES and
+                (self.method != b"CONNECT" and self.code/100 == 2)):
```
Note that you need `//` here on python 3.

mkzero committed:
Fixed modulus operator for Python3 compatibility

mkzero committed:
Only remove Content-Length header when response is 200

As pointed out by njsmith the absense of the header is only relevant for
2xx responses

wsanchez committed:
Merge branch 'trunk' into 9076-mkzero-t-w-web-connect-compatibility

wsanchez commented:
Test pass except for missing news fragment.  Instructions for those are here: https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles

izzy commented:
@wsanchez I added already one.. not sure why that's triggering as error...?

rodrigc commented:
@mkzero you submitted your change earlier.  After you submitted your change, unfortunately the process for newsfiles changes, *after* you submitted your change.
What you need to do is **git mv** one of your files from:
**src/twisted/web/topfiles/9076.bugfix** to **src/twisted/web/newsfragments/9076.bugfix**.

After you do that, commit, and push the change.

rodrigc commented:
Also, before  you push your change to this branch, can you merge latest trunk into your branch?
There have been a lot of changes.

mkzero committed:
Merge remote-tracking branch 'upstream/trunk' into 9076-mkzero-t-w-web-connect-compatibility

mkzero committed:
Fixed newsfile path for new format

mkzero committed:
Merge branch '9076-mkzero-t-w-web-connect-compatibility' of github.com:mkzero/twisted into 9076-mkzero-t-w-web-connect-compatibility

mkzero committed:
Added missing newline for txchecker

izzy commented:
So I fixed a few things merged trunk into my branch, but I'm not sure if the error remaining in [Travis](https://travis-ci.org/twisted/twisted/jobs/283736825#L3153) is my fault..?

rodrigc commented:
It's not your fault.  Travis started using Python 3.6.3, and something changed there causing test_sslverify.py to break

izzy commented:
@rodrigc so AFAICS my PR should be mergeable?

rodrigc commented:
This cannot be merged, because the failing Travis build will block this.
This is affecting all PR's, not just yours.

@mkzero Can you code review this: by clicking the green *Review changes* button at https://github.com/twisted/twisted/pull/910/files 

I need someone to code review that so I can merge it into trunk.  If I can get this fix into trunk, we can merge trunk into your branch, and get your branch passing Travis again.

rodrigc commented:
@mkzero I got someone else to code review https://github.com/twisted/twisted/pull/910 and it has been merged to trunk.  Can you merge Twisted trunk into your branch?  That will get Travis passing on your branch.

mkzero committed:
Merge remote-tracking branch 'upstream/trunk' into 9076-mkzero-t-w-web-connect-compatibility

izzy commented:
Merged trunk into my branch but it seems the buildbot check hangs?

rodrigc commented:
That's normal, since only a Twisted committer can trigger all the buildbots.

rodrigc commented on src/twisted/web/test/test_http.py:
```diff
@@ -3213,6 +3213,35 @@ def test_expect100ContinueHeader(self):
+    """
+
+    def test_headers(self):
+        transport = StringTransport()
```
Please add docstring here to eliminate error here: https://travis-ci.org/twisted/twisted/jobs/284299844.
When writing docstring, follow this guideline: https://jml.io/pages/test-docstrings.html
and don't put **Test** in the description.

mkzero committed:
Added missing docstring for ConnectRequestTests

mkzero committed:
Merge branch 'trunk' into 9076-mkzero-t-w-web-connect-compatibility

mkzero committed:
Merge branch 'trunk' into 9076-mkzero-t-w-web-connect-compatibility

izzy commented:
Seems like the tests finally are all green. Anything else holding it back from being merged into trunk?

Aurora-Beta commented:
I'm affected by the issue as well. Can anyone say when i can expect this to be merged?

mkzero committed:
Merge branch 'trunk' into 9076-mkzero-t-w-web-connect-compatibility

mkzero committed:
Merge branch 'trunk' into 9076-mkzero-t-w-web-connect-compatibility

mkzero committed:
Merge branch 'trunk' into 9076-mkzero-t-w-web-connect-compatibility

izzy commented:
Can someone trigger the buildbot and merge this if possible? I've been waiting almost a year to get this bugfix into Twisted now...

exarkun commented:
Triggered buildbot

exarkun commented:
The fedeora24-py3.5 failure is unrelated to this PR so it shouldn't block the merge.  It's the known issue - https://twistedmatrix.com/trac/ticket/9382

exarkun commented on src/twisted/web/test/test_http.py:
```diff
@@ -3325,6 +3325,41 @@ def test_expect100ContinueHeader(self):
+        channel.dataReceived(b"\r\n")
+
+        self.assertTrue(
+            transport.value().startswith(b"HTTP/1.1 200 OK\r\n"))
```
This assertion looks redundant with the following.  Is there a particular reason to have it?  If not, I think it should be removed.  The `assertResponseEquals` will cover this check.

exarkun commented on src/twisted/web/test/test_http.py:
```diff
@@ -3325,6 +3325,41 @@ def test_expect100ContinueHeader(self):
 
+class ConnectRequestTests(unittest.TestCase, ResponseTestMixin):
+    """
+    Test correct RFC 7231 handling for CONNECT requests. Responses to these
```
You can leave off "correct" here.  Tests are always for correct behavior. :wink:

exarkun commented on src/twisted/web/test/test_http.py:
```diff
@@ -3325,6 +3325,41 @@ def test_expect100ContinueHeader(self):
+
+    def test_headers(self):
+        """
+        A HTTP connection is opened and a CONNECT request is received, which
```
`An HTTP ...`

exarkun commented on src/twisted/web/test/test_http.py:
```diff
@@ -3325,6 +3325,41 @@ def test_expect100ContinueHeader(self):
+        A HTTP connection is opened and a CONNECT request is received, which
+        should return a 200 response with the HTTP version header, the host and
+        port as well as the CONNECT command but not a Transfer-Encoding or
+        Content-Length header, which both are not allowed in a 200 response.
```
`... neither of which is allowed in a 200 response.` would be more natural English.

exarkun commented on src/twisted/web/http.py:
```diff
@@ -1085,6 +1089,11 @@ def write(self, data):
                             category=DeprecationWarning, stacklevel=2)
                         # Backward compatible cast for non-bytes values
                         value = networkString('%s' % (value,))
+                    if (self.method == b"CONNECT" and self.code//100 == 2 and
```
There should be space surrounding the `//` operator.

exarkun commented on src/twisted/web/http.py:
```diff
@@ -1056,9 +1056,13 @@ def write(self, data):
                 (self.responseHeaders.getRawHeaders(b'content-length') is None) and
-                self.method != b"HEAD" and self.code not in NO_BODY_CODES):
+                self.method != b"HEAD" and self.code not in NO_BODY_CODES and
+                (self.method != b"CONNECT" and self.code//100 == 2)):
```
There should be space surrounding the `//` operator.

exarkun commented on src/twisted/web/newsfragments/9076.bugfix:
```diff
@@ -0,0 +1 @@
+twisted.web.http now handles CONNECT requests according to RFC 7231 and will not send Content-Length or Transfer-Encoding headers anymore
```
Please add a full stop (`.`) to the end of this sentence.

exarkun commented on src/twisted/web/http.py:
```diff
@@ -1085,6 +1089,11 @@ def write(self, data):
                         value = networkString('%s' % (value,))
+                    if (self.method == b"CONNECT" and self.code//100 == 2 and
+                        name == b"Content-Length"):
+                        log.msg("Warning: Removing Content-Length header in"
```
It looks like the logging style in this module has changed in the interim.  Can you make this `self._log.info(...)` instead (I would almost say something about not putting "Warning" there but I see that other code in this module emits info messages that start with "Warning" so ...)

exarkun commented on src/twisted/web/http.py:
```diff
@@ -1056,9 +1056,13 @@ def write(self, data):
             # chunked mode, so that we can support pipelining in
             # persistent connections.
+            # For no-body responses and CONNECT requests this
+            # does not apply and the header should not or must
```
is it "*should* not" or "*must* not"?  seems like it should be one or the other.

exarkun reviewed (approved):
Thanks.  Sorry about the long review delay.  I've left some minor comments inline.  I'd like to see them addressed but then I think this can be merged.

mkzero committed:
Minor style changes according to feedback in #741

mkzero committed:
Merge branch 'trunk' into 9076-mkzero-t-w-web-connect-compatibility

mkzero committed:
Merge branch 'trunk' into 9076-mkzero-t-w-web-connect-compatibility

izzy commented:
@exarkun I've implemented the changes, tests seem to run fine, can you run the buildbot again?

twm commented:
I merged this forward in #1358. Closing this one as the original fork has disappeared, so it can't go anywhere.

izzy commented:
Just for completeness sake, I didn't delete the fork per se, I guess Github just didn't handle me changing my username properly, sorry if that caused any inconveniences.

twm commented:
@izzy Thanks for clarifying!
